avocado/plugins/replay.py

```diff
@@ -0,0 +1,96 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat, Inc. 2020
+# Authors: Cleber Rosa <crosa@redhat.com>
+
+"""Replay Job Plugin"""
+
+import json
+import os
+import sys
+
+from avocado.core.data_dir import get_job_results_dir
+from avocado.core.dispatcher import JobPrePostDispatcher
+from avocado.core.future.settings import settings
+from avocado.core import job, exit_codes, output
+from avocado.core.plugin_interfaces import CLICmd
+
+
+class Replay(CLICmd):
+
+    """Implements the avocado 'replay' subcommand."""
+
+    name = 'replay'
+    description = "Runs a new job using a previous job as its configuration"
+
+
+    def configure(self, parser):
+        parser = super(Replay, self).configure(parser)
+
+        help_msg = ('Replays a job, identified by: complete or partial Job '
+                    'ID, "latest" for the latest job, the job results path.')
+        settings.register_option(section='job.replay',
+                                 key='source_job_id',
+                                 help_msg=help_msg,
+                                 metavar='SOURCE JOB ID',
+                                 default='latest',
+                                 nargs='?',
+                                 positional_arg=True,
+                                 parser=parser)
+
+    @staticmethod
+    def _get_source_job_results_dir(source_job_id):
+        results_dir = get_job_results_dir(source_job_id)
+        if results_dir:
+            return results_dir
+        msg = 'Could not find the results directory for Job "%s"'
+        output.LOG_UI.error(msg, source_job_id)
+        sys.exit(exit_codes.AVOCADO_FAIL)
+
+    @staticmethod
+    def _exit_config_file_not_found(config_path):
+        msg = 'Could not open the source Job configuration "%s"'
+        output.LOG_UI.error(msg, config_path)
+        sys.exit(exit_codes.AVOCADO_FAIL)
+
+    @staticmethod
+    def _exit_config_file_corrupt(config_path):
+        msg = 'Could not read a valid configuration from file "%s"'
+        output.LOG_UI.error(msg, config_path)
+        sys.exit(exit_codes.AVOCADO_FAIL)
+
+    @staticmethod
+    def _get_source_job_config(source_job_id):
+        results_dir = Replay._get_source_job_results_dir(source_job_id)
+        config_file_path = os.path.join(results_dir, 'jobdata/args.json')
+        try:
+            with open(config_file_path, 'r') as config_file:
+                return json.load(config_file)
+        except OSError:
+            Replay._exit_config_file_not_found(config_file_path)
+        except json.decoder.JSONDecodeError:
+            Replay._exit_config_file_corrupt(config_file_path)
+
+    def run(self, config):
+        source_job_id = config.get('job.replay.source_job_id')
+        source_job_config = self._get_source_job_config(source_job_id)
+        del(source_job_config['job.replay.source_job_id'])
+        with job.Job(source_job_config) as job_instance:
```

Comment:
According to our discussion and my PR about multiple test suites, Job() constructor will be used for "manual/custom" operations, because of that no magic will be assumed, so we need to pass `test_suites` here.

Reply:
Then I could change this for `job.Job.from_config(...)`, right?

Reply:
Yes, but outside the with, right?

Reply:
Why?  This works:

```python
#!/usr/bin/env python3
import sys

from avocado.core.job import Job

job_config = {'run.test_runner': 'nrunner',
              'run.references': ['examples/tests/passtest.py:PassTest.test']}

# Automatic helper method (Avocado will try to discovery things from config
# dicts. Since there is magic here, we dont need to pass suite names or suites,
# and test/task id will be prepend with the suite index (in this case 1 and 2)

with Job.from_config(job_config=job_config) as j:
    sys.exit(j.run())
```

Reply:
@clebergnu we have today an `__enter__()` method inside Job() that, afaik, was designed to hide some tasks when calling with/enter, right?

> The object's __enter__() is called before with-block is executed and therefore can run set-up code.

You can call `with` yes, but you need to run j.setup() and reconfigure, unless I'm missing something.

Reply:
No, you shouldn't need to call `setup`.  The context manager will still work, and it's separated from the instantiation.  This reminds that the changes you did to the example jobs are not really needed AFAICT.

Reply:
@clebergnu ok then, I will try to change there and keep the `with`.